File: crates/mycelium-config/src/topology.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ConfigError {
    #[error("Failed to read config file: {0}")]
    IoError(#[from] std::io::Error),

    #[error("Failed to parse config: {0}")]
    ParseError(#[from] toml::de::Error),

    #[error("Bundle not found: {0}")]
    BundleNotFound(String),

    #[error("Service not found: {0}")]
    ServiceNotFound(String),

    #[error("Invalid configuration: {0}")]
    InvalidConfig(String),
}

pub type Result<T> = std::result::Result<T, ConfigError>;

/// Top-level configuration for Mycelium deployment
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Topology {
    pub deployment: Deployment,

    #[serde(default)]
    pub bundles: Vec<Bundle>,

    #[serde(default)]
    pub inter_bundle: Option<InterBundleConfig>,
}

/// Deployment mode configuration
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Deployment {
    pub mode: DeploymentMode,
}

/// Deployment modes
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum DeploymentMode {
    /// All services in one process (Arc<T>)
    Monolith,

    /// Services grouped in bundles (Arc within, Unix/TCP between)
    Bundled,

    /// Each service in its own process (Unix/TCP all)
    Isolated,

    /// Services across machines (TCP)
    Distributed,
}

/// Bundle of services that run in the same process
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Bundle {
    pub name: String,
    pub services: Vec<String>,

    /// Optional: hostname for distributed deployment
    #[serde(default)]
    pub host: Option<String>,

    /// Optional: TCP port for distributed deployment
    #[serde(default)]
    pub port: Option<u16>,
}

/// Configuration for inter-bundle communication
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct InterBundleConfig {
    pub transport: TransportType,

    #[serde(default = "default_socket_dir")]
    pub socket_dir: PathBuf,
}

fn default_socket_dir() -> PathBuf {
    PathBuf::from("/tmp/mycelium")
}

/// Transport type for messages
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum TransportType {
    /// In-process Arc<T> (zero-copy)
    Local,

    /// Unix domain sockets (same machine)
    Unix,

    /// TCP sockets (network)
    Tcp,
}

impl Topology {
    /// Load topology from TOML file
    pub fn load<P: AsRef<Path>>(path: P) -> Result<Self> {
        let contents = fs::read_to_string(path)?;
        let topology: Topology = toml::from_str(&contents)?;
        topology.validate()?;
        Ok(topology)
    }

    /// Validate the topology configuration
    fn validate(&self) -> Result<()> {
        // Check for duplicate bundle names
        let mut bundle_names = std::collections::HashSet::new();
        for bundle in &self.bundles {
            if !bundle_names.insert(&bundle.name) {
                return Err(ConfigError::InvalidConfig(format!(
                    "Duplicate bundle name: {}",
                    bundle.name
                )));
            }
        }

        // Check for duplicate service names across bundles
        let mut service_names = std::collections::HashSet::new();
        for bundle in &self.bundles {
            for service in &bundle.services {
                if !service_names.insert(service) {
                    return Err(ConfigError::InvalidConfig(format!(
                        "Service {} appears in multiple bundles",
                        service
                    )));
                }
            }
        }

        // Validate distributed deployment config
        if self.deployment.mode == DeploymentMode::Distributed {
            for bundle in &self.bundles {
                if bundle.host.is_some() && bundle.port.is_none() {
                    return Err(ConfigError::InvalidConfig(format!(
                        "Bundle {} has host but no port",
                        bundle.name
                    )));
                }
                if bundle.host.is_none() && bundle.port.is_some() {
                    return Err(ConfigError::InvalidConfig(format!(
                        "Bundle {} has port but no host",
                        bundle.name
                    )));
                }
            }
        }

        Ok(())
    }

    /// Find which bundle a service belongs to
    pub fn find_bundle(&self, service_name: &str) -> Option<&Bundle> {
        self.bundles
            .iter()
            .find(|b| b.services.iter().any(|s| s == service_name))
    }

    /// Check if two services are in the same bundle
    pub fn same_bundle(&self, service1: &str, service2: &str) -> bool {
        if let Some(bundle1) = self.find_bundle(service1) {
            if let Some(bundle2) = self.find_bundle(service2) {
                return bundle1.name == bundle2.name;
            }
        }
        false
    }

    /// Get the transport type for communication between two services
    pub fn transport_between(&self, from_service: &str, to_service: &str) -> TransportType {
        match self.deployment.mode {
            DeploymentMode::Monolith => TransportType::Local,

            DeploymentMode::Bundled => {
                if self.same_bundle(from_service, to_service) {
                    TransportType::Local
                } else {
                    self.inter_bundle
                        .as_ref()
                        .map(|ib| ib.transport)
                        .unwrap_or(TransportType::Unix)
                }
            }

            DeploymentMode::Isolated => {
                self.inter_bundle
                    .as_ref()
                    .map(|ib| ib.transport)
                    .unwrap_or(TransportType::Unix)
            }

            DeploymentMode::Distributed => {
                // Check if services are on the same host
                if let (Some(bundle1), Some(bundle2)) = (
                    self.find_bundle(from_service),
                    self.find_bundle(to_service),
                ) {
                    match (&bundle1.host, &bundle2.host) {
                        (Some(h1), Some(h2)) if h1 == h2 => TransportType::Unix,
                        (None, None) => TransportType::Local,
                        _ => TransportType::Tcp,
                    }
                } else {
                    TransportType::Local
                }
            }
        }
    }

    /// Get the Unix socket path for a bundle
    pub fn socket_path(&self, bundle_name: &str) -> PathBuf {
        let socket_dir = self
            .inter_bundle
            .as_ref()
            .map(|ib| ib.socket_dir.clone())
            .unwrap_or_else(default_socket_dir);

        socket_dir.join(format!("{}.sock", bundle_name))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_monolith_config() {
        let toml = r#"
            [deployment]
            mode = "monolith"
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        assert_eq!(topology.deployment.mode, DeploymentMode::Monolith);
    }

    #[test]
    fn test_parse_bundled_config() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "adapters"
            services = ["polygon-adapter", "ethereum-adapter"]

            [[bundles]]
            name = "strategies"
            services = ["flash-arbitrage"]

            [inter_bundle]
            transport = "unix"
            socket_dir = "/tmp/mycelium"
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        assert_eq!(topology.deployment.mode, DeploymentMode::Bundled);
        assert_eq!(topology.bundles.len(), 2);
        assert_eq!(topology.bundles[0].name, "adapters");
        assert_eq!(topology.bundles[0].services.len(), 2);
    }

    #[test]
    fn test_find_bundle() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "adapters"
            services = ["polygon-adapter", "ethereum-adapter"]
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        let bundle = topology.find_bundle("polygon-adapter").unwrap();
        assert_eq!(bundle.name, "adapters");
    }

    #[test]
    fn test_same_bundle() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "adapters"
            services = ["polygon-adapter", "ethereum-adapter"]

            [[bundles]]
            name = "strategies"
            services = ["flash-arbitrage"]
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        assert!(topology.same_bundle("polygon-adapter", "ethereum-adapter"));
        assert!(!topology.same_bundle("polygon-adapter", "flash-arbitrage"));
    }

    #[test]
    fn test_transport_between_bundled() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "adapters"
            services = ["polygon-adapter", "ethereum-adapter"]

            [[bundles]]
            name = "strategies"
            services = ["flash-arbitrage"]

            [inter_bundle]
            transport = "unix"
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();

        // Same bundle -> local
        assert_eq!(
            topology.transport_between("polygon-adapter", "ethereum-adapter"),
            TransportType::Local
        );

        // Different bundles -> unix
        assert_eq!(
            topology.transport_between("polygon-adapter", "flash-arbitrage"),
            TransportType::Unix
        );
    }

    #[test]
    fn test_socket_path() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "adapters"
            services = ["polygon-adapter"]

            [inter_bundle]
            transport = "unix"
            socket_dir = "/tmp/test"
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        let path = topology.socket_path("adapters");
        assert_eq!(path, PathBuf::from("/tmp/test/adapters.sock"));
    }

    #[test]
    fn test_validate_duplicate_bundle_names() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "test"
            services = ["service1"]

            [[bundles]]
            name = "test"
            services = ["service2"]
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        assert!(topology.validate().is_err());
    }

    #[test]
    fn test_validate_duplicate_service_names() {
        let toml = r#"
            [deployment]
            mode = "bundled"

            [[bundles]]
            name = "bundle1"
            services = ["service1"]

            [[bundles]]
            name = "bundle2"
            services = ["service1"]
        "#;

        let topology: Topology = toml::from_str(toml).unwrap();
        assert!(topology.validate().is_err());
    }
}

File: python-sdk/mycelium/topology.py
"""Topology configuration for Mycelium Python SDK."""

from __future__ import annotations

import sys
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Dict, List, Optional

# Use tomllib (Python 3.11+) or fallback to tomli
if sys.version_info >= (3, 11):
    import tomllib
else:
    try:
        import tomli as tomllib
    except ImportError:
        raise ImportError(
            "Python <3.11 requires 'tomli' package. "
            "Install with: pip install tomli"
        )


class TransportType(Enum):
    """Transport type for message delivery."""

    LOCAL = "local"  # In-process Arc<T> (not available for Python)
    UNIX = "unix"  # Unix domain sockets
    TCP = "tcp"  # TCP sockets


class EndpointKind(Enum):
    """Endpoint kind for socket endpoints."""

    TCP = "tcp"
    UNIX = "unix"


@dataclass
class EndpointConfig:
    """Configuration for exposing a node via socket endpoint."""

    kind: EndpointKind
    addr: Optional[str] = None


@dataclass
class Node:
    """Node: a process that runs one or more services.

    Transport is inferred from topology:
    - Services in same node → Local (Arc<T>, Rust only)
    - Nodes on same host → Unix sockets
    - Nodes on different hosts → TCP
    """

    name: str
    services: List[str]
    host: Optional[str] = None
    port: Optional[int] = None
    endpoint: Optional[EndpointConfig] = None


@dataclass
class Topology:
    """Top-level configuration for Mycelium deployment.

    Supports loading from TOML files and querying transport types
    between services based on topology.
    """

    nodes: List[Node]
    socket_dir: Path = Path("/tmp/mycelium")

    @classmethod
    def load(cls, path: str | Path) -> Topology:
        """Load topology from TOML file.

        Args:
            path: Path to topology.toml file

        Returns:
            Parsed and validated Topology

        Raises:
            ValueError: If topology is invalid
            FileNotFoundError: If file doesn't exist
        """
        path = Path(path)
        with open(path, "rb") as f:
            data = tomllib.load(f)

        nodes = []
        for node_data in data.get("nodes", []):
            endpoint_data = node_data.get("endpoint")
            endpoint = None
            if endpoint_data:
                kind = EndpointKind(endpoint_data["kind"])
                endpoint = EndpointConfig(
                    kind=kind, addr=endpoint_data.get("addr")
                )

            nodes.append(
                Node(
                    name=node_data["name"],
                    services=node_data["services"],
                    host=node_data.get("host"),
                    port=node_data.get("port"),
                    endpoint=endpoint,
                )
            )

        socket_dir = Path(data.get("socket_dir", "/tmp/mycelium"))
        topology = cls(nodes=nodes, socket_dir=socket_dir)
        topology.validate()
        return topology

    def validate(self) -> None:
        """Validate topology configuration.

        Raises:
            ValueError: If topology is invalid
        """
        # Check for duplicate node names
        node_names = set()
        for node in self.nodes:
            if node.name in node_names:
                raise ValueError(f"Duplicate node name: {node.name}")
            node_names.add(node.name)

        # Check for duplicate service names across nodes
        service_names = set()
        for node in self.nodes:
            for service in node.services:
                if service in service_names:
                    raise ValueError(
                        f"Service {service} appears in multiple nodes"
                    )
                service_names.add(service)

        # Validate nodes with remote hosts have ports
        for node in self.nodes:
            if node.host and node.host not in ("localhost", "127.0.0.1"):
                if node.port is None:
                    raise ValueError(
                        f"Node {node.name} has remote host "
                        f"'{node.host}' but no port specified"
                    )

    def find_node(self, service_name: str) -> Optional[Node]:
        """Find which node a service belongs to.

        Args:
            service_name: Name of the service to find

        Returns:
            Node containing the service, or None if not found
        """
        for node in self.nodes:
            if service_name in node.services:
                return node
        return None

    def same_node(self, service1: str, service2: str) -> bool:
        """Check if two services are in the same node.

        Args:
            service1: First service name
            service2: Second service name

        Returns:
            True if both services are in the same node
        """
        node1 = self.find_node(service1)
        node2 = self.find_node(service2)
        if node1 and node2:
            return node1.name == node2.name
        return False

    def transport_between(
        self, from_service: str, to_service: str
    ) -> TransportType:
        """Get the transport type for communication between two services.

        Transport is inferred from topology:
        - Same node → Local (Arc<T>, not available for Python)
        - Different nodes, same host → Unix sockets
        - Different nodes, different hosts → TCP

        Args:
            from_service: Source service name
            to_service: Target service name

        Returns:
            Transport type to use for communication
        """
        # Same node → Local (not usable from Python)
        if self.same_node(from_service, to_service):
            return TransportType.LOCAL

        # Different nodes → check hosts
        node1 = self.find_node(from_service)
        node2 = self.find_node(to_service)

        if node1 and node2:
            h1 = node1.host
            h2 = node2.host

            # Both on localhost or both unspecified → Unix
            if h1 is None and h2 is None:
                return TransportType.UNIX

            localhost_vals = ("localhost", "127.0.0.1")
            if (h1 in localhost_vals and h2 is None) or (
                h1 is None and h2 in localhost_vals
            ):
                return TransportType.UNIX

            if h1 == h2:
                return TransportType.UNIX

            # Different hosts → TCP
            return TransportType.TCP

        # Fallback
        return TransportType.LOCAL

    def socket_path(self, node_name: str) -> Path:
        """Get the Unix socket path for a node.

        Args:
            node_name: Name of the node

        Returns:
            Path to Unix socket for this node
        """
        return self.socket_dir / f"{node_name}.sock"

    def connection_info(self, target_service: str) -> Dict[str, any]:
        """Get connection information for connecting to a target service.

        Returns a dictionary with transport type and connection details.

        Args:
            target_service: Name of service to connect to

        Returns:
            Dict with keys:
            - 'transport': TransportType enum value
            - 'path': Unix socket path (if Unix transport)
            - 'host': TCP host (if TCP transport)
            - 'port': TCP port (if TCP transport)

        Raises:
            ValueError: If service not found in topology
        """
        node = self.find_node(target_service)
        if not node:
            raise ValueError(f"Service '{target_service}' not found in topology")

        # Determine transport type based on node configuration
        if node.host and node.host not in ("localhost", "127.0.0.1"):
            # Remote node → TCP
            return {
                "transport": TransportType.TCP,
                "host": node.host,
                "port": node.port,
            }
        else:
            # Local node → Unix socket
            return {
                "transport": TransportType.UNIX,
                "path": str(self.socket_path(node.name)),
            }


__all__ = [
    "Topology",
    "Node",
    "TransportType",
    "EndpointKind",
    "EndpointConfig",
]
